extensions/grpc_logging/plugin.cc

```diff
@@ -0,0 +1,182 @@
+    LOG_DEBUG("successfully sent loggin request");
+    in_flight_export_call_ -= 1;
+    if (in_flight_export_call_ < 0) {
+      LOG_WARN("in flight report call should not be negative");
+    }
+    if (in_flight_export_call_ <= 0 && is_on_done_) {
+      proxy_done();
+    }
+  };
+
+  failure_callback_ = [this](GrpcStatus status) {
+    LOG_WARN(absl::StrCat(
+        "Logging call error: ", std::to_string(static_cast<int>(status)),
+        getStatus().second->toString()));
+    in_flight_export_call_ -= 1;
+    if (in_flight_export_call_ < 0) {
+      LOG_WARN("in flight report call should not be negative");
+    }
+    if (in_flight_export_call_ <= 0 && is_on_done_) {
+      proxy_done();
+    }
+  };
+
+  // Start timer, which will trigger log report every 10s.
+  proxy_set_tick_period_milliseconds(10000 /* milliseconds */);
+
+  return true;
+}
+
+bool PluginRootContext::onDone() {
+  if (req_buffer_.empty() && in_flight_export_call_ == 0) {
+    return true;
+  }
+  // Flush out all log entries
+  flushLogBuffer();
+  sendLogRequest(/* ondone */ true);
+  return false;
+}
+
+void PluginRootContext::onTick() {
+  // Flush out all log entries
+  flushLogBuffer();
+  if (req_buffer_.empty()) {
+    return;
+  }
+  sendLogRequest(/* ondone */ false);
+}
+
+void PluginRootContext::addLogEntry(PluginContext* stream) {
+  auto* log_entries = cur_log_req_->mutable_log_entries();
+  auto* new_entry = log_entries->Add();
+
+  // Add log labels. Note the following logic assumes this extension
+  // is running at a server sidecar.
+
+  // Workload attributes.
+  std::string source_address, destination_address;
+  getValue({"source", "address"}, &source_address);
+  getValue({"destination", "address"}, &destination_address);
+  new_entry->set_source_address(source_address);
+  new_entry->set_destination_address(destination_address);
+
+  std::string destination_workload, destination_namespace;
+  getValue({"node", "metadata", "WORKLOAD_NAME"}, &destination_workload);
+  getValue({"node", "metadata", "NAMESPACE"}, &destination_namespace);
+  new_entry->set_destination_workload(destination_workload);
+  new_entry->set_destination_namespace(destination_namespace);
+
+  // Request attributes.
+  std::string path, host, request_id, referer, user_agent;
+  int64_t response_code, timestamp, duration;
+  getValue({"request", "time"}, &timestamp);
+  getValue({"request", "duration"}, &duration);
+  getValue({"request", "id"}, &request_id);
+  getValue({"request", "host"}, &host);
+  getValue({"request", "url_path"}, &path);
+  getValue({"response", "code"}, &response_code);
+  getValue({"request", "referer"}, &referer);
+  getValue({"request", "user_agent"}, &user_agent);
+
+  *new_entry->mutable_timestamp() =
+      google::protobuf::util::TimeUtil::NanosecondsToTimestamp(timestamp);
+  *new_entry->mutable_latency() =
+      google::protobuf::util::TimeUtil::NanosecondsToDuration(duration);
+  new_entry->set_request_id(request_id);
```

Comment:
You could just have passed `mutable_request_id` directly to `getValue`.

Reply:
Much better! Thanks!